File: CPP/Second/number_of_islands.cc
#include <iostream>
#include <algorithm>
#include <vector>

using namespace std;

class Solution {
public:
    int numIslands(vector<vector<char>> &grid) {
        int cnt = 0;
        if (grid.size() < 1)
            return 0;
        int m = grid.size(), n = grid[0].size();
        vector<vector<bool> > mark(m, vector<bool>(n, false));
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (dfs(i, j, grid, mark)) {
                    cnt++;
                }
            }
        }
        return cnt;
    }
    bool dfs(int i,int j, vector<vector<char> > &grid, vector<vector<bool> > &mark) {
        if (grid[i][j] == '0' || mark[i][j] == true)
            return false;
        mark[i][j] = true;
        if (i-1 >= 0)
            dfs(i-1, j, grid, mark);
        if (i+1 < grid.size())
            dfs(i+1, j, grid, mark);
        if (j-1 >= 0)
            dfs(i, j-1, grid, mark);
        if (j+1 < grid[0].size())
            dfs(i, j+1, grid, mark);
        return true;
    }
};

File: NumberOfIslands.py
class Solution:
    # @param {character[][]} grid
    # @return {integer}
    def numIslands(self, grid):
        if len(grid) < 1:
            return 0
        height, width = len(grid), len(grid[0])
        mark = [[False for y in range(width)] for x in range(height)]
        res = 0
        for i in range(height):
            for j in range(width):
                if self.dfs(i, j, grid, mark):
                    res += 1
        return res

    def dfs(self, i, j, grid, mark):
        """
        Args:
            i, j: int value to mark coord
            mark: 2D Boolean matrix to mark visited
        Returns:
            True if this is a new island
        """
        if i < 0 or j < 0 or i >= len(mark) or j >= len(mark[0]):
            return False
        if grid[i][j] == '0' or mark[i][j]:
            return False
        mark[i][j] = True
        self.dfs(i - 1, j, grid, mark)
        self.dfs(i + 1, j, grid, mark)
        self.dfs(i, j - 1, grid, mark)
        self.dfs(i, j + 1, grid, mark)
        return True
